### PR TITLE
Implement PushToNamedFork in the git v2 client.

### PR DESCRIPTION
**Pull request by cjwagner**
I think this is a reasonable way to implement this, but I'm definitely open to alternative implementation suggestions. I tried to refactor the logic for the `httpResolverFactory` into something more readable.

fixes https://github.com/kubernetes/test-infra/issues/30937

/assign @airbornepony @timwangmusic 
/cc @killianmuldoon

### CODE REVIEW

File: prow/git/v2/remote_test.go [hunk 1]
```diff
@@ -19,75 +19,10 @@ package git
...
-			username: func() (string, error) {
-				return "", errors.New("oops")
-			},
-			expectedErr: true,
-		},
-	}
-
-	for _, testCase := range testCases {
-		t.Run(testCase.name, func(t *testing.T) {
-			actual, actualErr := HttpResolver(testCase.remote, testCase.username, testCase.token, "org")()
```
Comment (cjwagner) [inline prow/git/v2/remote_test.go:77]: The functionality that was tested here is well tested by the other test functions in this file. I updated the other test functions with unit cases to cover the new behavior.

File: prow/git/v2/remote.go [hunk 2]
```diff
@@ -65,8 +70,11 @@ func (f *sshRemoteResolverFactory) CentralRemote(org, repo string) RemoteResolve
 
 // PublishRemote creates a remote resolver that refers to a user's remote
 // for the repository that can be published to.
-func (f *sshRemoteResolverFactory) PublishRemote(_, repo string) RemoteResolver {
-	return func() (string, error) {
+func (f *sshRemoteResolverFactory) PublishRemote(_, repo string) ForkRemoteResolver {
+	return func(forkName string) (string, error) {
+		if forkName != "" {
+			repo = forkName
```
Comment (airbornepony) [inline prow/git/v2/remote.go:76]: I believe this assignment persists, i.e. repo keeps the last non-empty forkName value passed. Think you need to create a local "r := repo" and optionally reassign that if you didn't intend for it to be persistent.

File: prow/git/v2/remote.go [hunk 3]
```diff
@@ -87,56 +95,58 @@ type httpResolverFactory struct {
...
-		scheme := "https"
-		if f.http {
-			scheme = "http"
+func (f *httpResolverFactory) PublishRemote(_, repo string) ForkRemoteResolver {
+	return func(forkName string) (string, error) {
+		// For the publsh remote we use:
+		// - the user login rather than the upstream org
+		// - the forkName rather than the upstream repo name, if specified.
+		if forkName != "" {
+			repo = forkName
```
Comment (airbornepony) [inline prow/git/v2/remote.go:111]: Same here, repo assignment is persistent.

File: prow/git/v2/remote.go (hunk 2 above)
Comment (cjwagner) [reply prow/git/v2/remote.go:77]: Nice catch, that's subtle.
I fixed it, changed the variable names to distinguish which repo we're actually referring to, and added unit test coverage for this bug.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Comment by MartinForReal** [comment]
I'm wondering if we can deploy the prow changes ASAP because cherrypicker is a frequently used utilility. @k8s-infra-group-admins

**Pull request merged**